Shared/PrivilegedRequest.swift

```diff
@@ -0,0 +1,98 @@
+
+import Foundation
+
+private let REQUEST_KEY_PRIVILEGED = "privileged"
+
+/**
+ Request that is allowed to load local resources.
+
+ Pages running on the local server have same origin access all resources
+ on the server, so we need to prevent arbitrary web pages from accessing
+ these resources. We do so by explicitly requiring "privileged" requests
+ in our navigation policy when loading local resources.
+
+ Be careful: creating a privileged request for an arbitrary URL provided
+ by the page will break this model. Only use a privileged request when
+ needed, and when you are sure the URL is from a trustworthy source!
+ **/
+public class PrivilegedRequest: NSMutableURLRequest {
+    public static let key = "brave_prv"
+    public static let token = UUID().uuidString.replacingOccurrences(of: "-", with: "")
+
+    override init(url URL: URL, cachePolicy: NSURLRequest.CachePolicy, timeoutInterval: TimeInterval) {
+        let modifyURL = { (url: URL) -> URL in
+            if PrivilegedRequest.isWebServerRequest(url: url),
+               let url = PrivilegedRequest.store(url: url) {
+                return url
+            }
+            return url
+        }
+
+        super.init(url: modifyURL(URL), cachePolicy: cachePolicy, timeoutInterval: timeoutInterval)
+        setPrivileged()
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+        setPrivileged()
+    }
+
+    private func setPrivileged() {
+        URLProtocol.setProperty(true, forKey: REQUEST_KEY_PRIVILEGED, in: self)
+    }
+
+    private static func store(url: URL) -> URL? {
+        guard var components = URLComponents(string: url.absoluteString) else { return nil }
+
+        var queryItems = components.queryItems ?? []
+        if var item = queryItems.find({ $0.name == PrivilegedRequest.key }) {
+            item.value = PrivilegedRequest.token
+        } else {
+            let queryItem = URLQueryItem(name: PrivilegedRequest.key,
+                                         value: PrivilegedRequest.token)
+            queryItems.append(queryItem)
+        }
+
+        components.queryItems = queryItems
+        return components.url
+    }
+
+    public static func removePrivileges(url: URL) -> URL? {
+        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false),
+                let items = components.queryItems else { return url }
+
+        components.queryItems = items.filter { $0.name != PrivilegedRequest.key }
+        if let items = components.queryItems,
+            items.isEmpty {
+            components.queryItems = nil
+        }
+        return components.url
+    }
+
+    public static func isPrivileged(url: URL?) -> Bool {
+        if let value = url?.getQuery()[PrivilegedRequest.key],
+           !value.isEmpty,
+           value == PrivilegedRequest.token {
+            return true
+        }
+        return false
+    }
+
+    public static func isWebServerRequest(url: URL) -> Bool {
```

Comment:
nit: This can be used in `init` and `isPrivileged` to avoid code duplication.

Reply:
It is used in `init` but I didn't put it in `isPrivileged` because I wasn't sure 100% if we have any non-localhost requests that are possibly `privileged`.

Reply:
All current accesses to `isPrivileged` are for localhost URLs. But, I think we can leave it as is for now and may be re-work this later.